dali/util/shared_mem.h

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_UTIL_SHARED_MEM_H_
+#define DALI_UTIL_SHARED_MEM_H_
+
+#include <stdint.h>
+#include <memory>
+#include "dali/core/common.h"
+
+namespace dali {
+namespace python {
+
+/**
+ * @brief Creates or simply stores (if provided) file descriptor used for identifying
+ * shared memory chunk. Closes file descriptor in the desctructor if it
+ * wasn't closed earlier.
+ */
+class DLL_PUBLIC ShmFdWrapper {
+ public:
+  /**
+   * @brief Wrap a file descriptor, fd shouldn't be -1.
+   */
+  explicit ShmFdWrapper(int fd);
+
+  /**
+   * Create new file descriptor that can be used for shared memory chunk.
+   */
+  ShmFdWrapper();
+
+  ~ShmFdWrapper();
+
+  DISABLE_COPY_MOVE_ASSIGN(ShmFdWrapper);
+
+  int get_fd();
+
+  int close();
+
+ private:
+  int fd_;
+};
+
+
+/**
+ * @brief Map/unmap and resize mapping of shared memory into process address space.
+ * Unmaps memory in the destructor if it wasn't unmapped earlier.
+ */
+class DLL_PUBLIC MapMemWrapper {
```

Comment:
1. I think this (and some file utilities, too) belong to some kind of OS sub-library (in core)?
2. Probably we should think of other platforms (Windows?) and try to abstract a bit from the Unix-centric model. This is basically a C++ wrapper for Unix shared memory objects, with no extra abstraction.

I think that at least we could use some typedefs for file/shm handles instead of directly using `int` and calling it `fd`.

Reply:
Also: What's MapMem? Isn't it a FileMapping or MappedMemChunk or sth along these lines?

Reply:
Adjusted the name to MemoryMapping. Used unique_handle. A bit of abstraction added (mostly by obscuring the fact that we are indeed using file descriptors). I tried my best but I'm no Windows programmer, so maybe any further generalizations should be made in a followup.

Reply:
Should I move it to /core/os? I think that maybe putting it as an utility and not a first-class API part has some merit.